File: src/lanefollower/detector.cpp
#include "rclcpp/rclcpp.hpp"
#include "sensor_msgs/msg/image.hpp"
#include "cv_bridge/cv_bridge.h"
#include "opencv2/opencv.hpp"

class ImageConverter : public rclcpp::Node
{
public:
    ImageConverter() : Node("detector")
    {
        // Subscribe to the RGB image topic
        subscription_ = this->create_subscription<sensor_msgs::msg::Image>(
            "/zed2i/zed_node/image_raw", 10,
            [this](const sensor_msgs::msg::Image::SharedPtr msg) {
                processImage(msg);
            });

        // Publish the grayscale image
        publisher_ = this->create_publisher<sensor_msgs::msg::Image>("/gray_image_topic", 10);
    }

private:
    void processImage(const sensor_msgs::msg::Image::SharedPtr msg)
    {
        try
        {
            // Convert ROS image message to OpenCV image
            cv_bridge::CvImagePtr cv_ptr = cv_bridge::toCvCopy(msg, sensor_msgs::image_encodings::RGB8);

            // Convert RGB image to grayscale
            cv::Mat gray_image;
            cv::cvtColor(cv_ptr->image, gray_image, cv::COLOR_RGB2GRAY);
            cv::inRange(gray_image, cv::Scalar(120), cv::Scalar(140), gray_image);
            // gray_image = gray_image(cv::Rect(0, gray_image.rows / 3 * 2, gray_image.cols, gray_image.rows/3));
            // Convert OpenCV grayscale image to ROS image message
            sensor_msgs::msg::Image::SharedPtr gray_msg = cv_bridge::CvImage(std_msgs::msg::Header(), "mono8", gray_image).toImageMsg();

            // Publish the grayscale image
            publisher_->publish(*gray_msg);
        }
        catch (cv_bridge::Exception& e)
        {
            RCLCPP_ERROR(this->get_logger(), "cv_bridge exception: %s", e.what());
        }
    }
    rclcpp::Subscription<sensor_msgs::msg::Image>::SharedPtr subscription_;
    rclcpp::Publisher<sensor_msgs::msg::Image>::SharedPtr publisher_;
};

int main(int argc, char *argv[])
{
    rclcpp::init(argc, argv);
    rclcpp::spin(std::make_shared<ImageConverter>());
    rclcpp::shutdown();
    return 0;
}

File: src/lanefollower/detector.py
#!/usr/bin/python3
import rclpy
from rclpy.node import Node
from sensor_msgs.msg import Image
from cv_bridge import CvBridge
import cv2
from sklearn.cluster import DBSCAN
import numpy as np

class ImageConverter(Node):
    def __init__(self):
        super().__init__('detector')

        # Subscribe to the RGB image topic
        self.subscription = self.create_subscription(
            Image,
            '/zed/zed_node/rgb/image_rect_color',
            self.process_image,
            10)

        # Publish the grayscale image
        self.publisher_ = self.create_publisher(Image, '/igvc/lanes_binary', 10)

    def process_image(self, msg):
        try:
            # Convert ROS image message to OpenCV image
            bridge = CvBridge()
            cv_image = bridge.imgmsg_to_cv2(msg, desired_encoding='passthrough')

            # Convert RGB image to grayscale
            gray_image = cv2.cvtColor(cv_image, cv2.COLOR_RGB2GRAY)
            binary_image = cv2.inRange(gray_image, 125, 140)
            binary_image[:binary_image.shape[0]//2 + 100, :] = 0

            # Find the indices of white pixels
            white_pixel_indices = np.argwhere(binary_image == 255)

            # Run DBSCAN on white pixel indices
            db = DBSCAN(eps=1, min_samples=5).fit(white_pixel_indices)
            labels = db.labels_
            no_clusters = len(np.unique(labels))
            no_noise = np.sum(labels == -1)

            print('Estimated no. of clusters: %d' % no_clusters)
            print('Estimated no. of noise points: %d' % no_noise)

            # Determine the size of each cluster
            cluster_sizes = [np.sum(labels == label) for label in np.unique(labels) if label != -1]

            # Keep the indices of the largest two clusters
            largest_clusters_indices = np.argsort(cluster_sizes)[-2:]

            # Create a blank image to draw clusters
            clustered_image = np.zeros_like(cv_image)

            for label in np.unique(labels):
                
                if label == -1 or label not in largest_clusters_indices:
                    continue
                
                cluster_indices = white_pixel_indices[labels == label]
                for point in cluster_indices:
                    clustered_image[point[0], point[1]] = 255

            clustered_msg = bridge.cv2_to_imgmsg(clustered_image, encoding='rgb8')


            # Publish the clustered image
            self.publisher_.publish(clustered_msg)
        except Exception as e:
            self.get_logger().error('cv_bridge exception: %s' % e)

def main(args=None):
    rclpy.init(args=args)
    image_converter = ImageConverter()
    rclpy.spin(image_converter)
    image_converter.destroy_node()
    rclpy.shutdown()

if __name__ == '__main__':
    main()
